db/hints/manager.cc

```diff
@@ -569,4 +576,85 @@ future<> manager::with_file_update_mutex_for(endpoint_id ep, noncopyable_functio
+        // If there is no mapping from an IP to a host ID, drop a directory by effectively removing it
+        // and its contents. Hinted Handoff is a "best-effort" mechanism -- there is no reason to deal
+        // with such edge cases, especially because this can only happen once -- during an upgrade.
+        co_await lister::scan_dir(_hints_dir, lister::dir_entry_types::of<directory_entry_type::directory>(),
+                [tmptr] (fs::path datadir, directory_entry de) -> future<> {
+            auto remove_directory = [] (fs::path p, const std::string_view reason) -> future<> {
+                // We cannot map the IP to the corresponding host ID either because
+                // the relevant mapping doesn't exist anymore or an error occurred. Drop it.
+                //
+                // We only care about directories named after IPs during an upgrade.
+                // In normal circumstances, this whole lambda should be equivalent to a no-op,
+                // i.e. it should have no side effects.
+                manager_logger.warn("Removing hint directory {}. Reason: {}", p.filename(), reason);
+                return lister::rmdir(p);
+            };
+
+            auto parse_name = [] (const sstring& name) -> std::optional<locator::host_id_or_endpoint> {
+                try {
+                    return locator::host_id_or_endpoint{name};
+                } catch (...) {
+                    return {};
+                }
+            };
+
+            // Old path: `<...>/<shard id>/<IP>`.
+            const fs::path current_path = datadir / de.name;
+            auto maybe_parsed_name = parse_name(de.name);
+
+            if (!maybe_parsed_name) {
+                return remove_directory(current_path,
+                        seastar::format("The directory's name {} is neither IP, nor host ID.", de.name));
+            }
+            
+            if (maybe_parsed_name.value().has_host_id()) {
+                return make_ready_future<>();
+            }
+
+            try {
+                // This either finds the corresponding host ID or throws an exception.
+                maybe_parsed_name.value().resolve(*tmptr);
+            } catch (...) {
+                return remove_directory(current_path, "Not able to map the IP to a host ID.");
+            }
+
+            // New path: `<...>/<shard id>/<host ID>`.
+            const fs::path hid_path = datadir / maybe_parsed_name.value().id.to_sstring();
+            
+            // Try to rename the hint directory. If renaming fails, we don't want to attempt it again
+            // or leave the directory. Simply remove it.
+            return rename_file(current_path.native(), hid_path.native()).then([current_path, hid_path] {
+                manager_logger.info("Successfully renamed directory {} to {}", current_path.filename(), hid_path.filename());
+            }).handle_exception([remove_directory, current_path] (auto&& e) {
```

Comment:
It's better to use coroutines in new code, they are more readable than `then/handle_exception/etc`

Reply:
We can't use `co_await` in `catch` clauses, but I'll see if we can use more C++20 code after changing the migration phases